[새기능] 날짜/시간 검증

cabbage16 opened the pull request:
## 🎫 관련 이슈
[//]: # (다음 키워드를 사용하면 해당 PR을 머지할 때 자동으로 이슈를 닫을 수 있습니다.)
[//]: # (keyword: close|closes|closed|resolve|resolves|resolved|fix|fixes|fixed)
[//]: # (예시: close #1)

close #69 

<br>

## 📄 개요
[//]: # (작업 내용을 간단히 요약해서 적습니다.)
[//]: # (예시: 유저 회원가입 기능을 만들었습니다.)

> 입학 일정에 맞춰 날짜 및 시간 검증 기능을 만들었습니다.

<br>

## 🔨 작업 내용
[//]: # (작업 내용을 자세하게 적습니다.)
[//]: # (붙임표 "-" 을 사용해서 목록을 만듭니다.)
[//]: # (예시: 유저 회원가입 API를 만들었습니다.)

- 스케줄을 상수에서 환경변수로 바꿔서 개발환경에서 편의성을 높였습니다.
- 원서 제출, 최종 제출, 수정, 증명사진 및 서류 업로드 시 원서 접수 기간이 아니라면 에러가 발생하도록 했습니다.
- 입학등록원 및 금연서약서 업로드 시 제출 기간이 아니라면 에러가 발생하도록 했습니다.


<br>

## 🏁 확인 사항
[//]: # (PR을 보내기 전 다음 사항을 확인해주세요.)
[//]: # (해당 사항을 모두 이행해야 머지할 수 있습니다.)
[//]: # (- [x] 를 사용해서 완료로 표시할 수 있습니다.)

- [x] 테스트를 완료했나요?
- [x] API 문서를 작성했나요?
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 로그, 주석, import 등을 삭제했나요?

<br>

## 🙋🏻 덧붙일 말
[//]: # (다음 사항이 있다면 적어주세요.)
[//]: # (PR에 대한 추가 설명)
[//]: # (중점적으로 리뷰받고 싶은 부분)
[//]: # (기타 등등)

스케줄을 환경변수로 바꾸는 과정에서 테스트에서 너무 많은 given과 verify 절이 생겼는데 어떻게 처리해야할 지 잘 모르겠어요.

cabbage16 committed:
feat(#69): 스케줄을 환경변수로 전환
- 개발 환경과 배포 환경에서 서로 스케줄이 달라야하므로 상수 대신 환경변수로 바꿨어요.
- Schedule이 ScheduleProperties에 의존하므로 유틸리티 클래스에서 서비스 클래스로 바꿨어요.

cabbage16 committed:
feat(#69): 원서 제출 시 날짜 및 시간 검증
- 원서를 제출, 최종 제출, 수정, 증명 사진 업로드, 원서 파일 업로드할 시 원서 접수 기간이 아니면 예외가 발생하도록 했어요.

cabbage16 committed:
refactor(#69): 원서 접수 기간이 아닐 시 에러 이름 변경
- 입학 등록원 접수 기간이 아닐때와 차이를 두기 위해서 예외 이름을 수정했어요.

cabbage16 committed:
feat(#69): 입학 등록원 및 서약서 제출 시 날짜 및 시간 검증
- 입학 등록원 및 서약서를 제출할 시 제출 기간이 아니면 예외가 발생하도록 했어요.

cabbage16 committed:
refactor(#69): 원서 제출 기간이 아닐 시 예외 이름 변경
- 원서 제출 기간이 아닐 시 예외 이름에 오타가 있어서 수정했어요.

cabbage16 committed:
test(#69): 원서 제출, 입학등록원 및 금연서약서 제출 시 날짜 및 시간 검증 테스트 코드 작성
- 원서를 제출, 최종 제출, 수정, 증명 사진 업로드, 서류 업로드 시 원서 접수 기간이 아니면 에러가 발생하는 테스트 코드를 작성했어요.
- 입학 등록원 및 금연 서약서를 제출할 시 제출 기간이 아니면 에러가 발생하는 테스트 코드를 작성했어요.

cabbage16 committed:
test(#69): 원서 제출, 입학등록원 및 금연서약서 제출 시 날짜 및 시간 검증 UseCaseTest 작성
- application-test.yml에 SCHEDULE 환경변수를 적용했어요.
- 원서 제출, 최종 제출, 수정, 증명사진 및 서류 업로드, 입학등록원 및 금연서약서 업로드 시 날짜 및 시간을 검증하는 UseCaseTest코드를 작성했어요.

cabbage16 committed:
test(#69): 스케줄 환경변수 관리에 따른 테스트 변경
- 스케줄을 상수에서 환경변수로 ScheduleProperties에서 관리하므로 테스트 코드를 수정했어요.

cabbage16 committed:
docs(#69): 날짜 및 시간 검증 API 문서 작성
- 원서 제출, 최종 제출, 수정, 증명 사진 및 서류 업로드, 입학등록원 및 금연서약서 업로드 시 정해진 기간이 아닐 경우 에러가 발생하는 내용을 작성했어요.

cabbage16 committed:
Merge branch 'develop' of https://github.com/Bamdoliro/marubase into feat/#69

cabbage16 committed:
Update gradle.yml

cabbage16 committed:
Update gradle.yml

cabbage16 committed:
Update gradle.yml

jyj1289 commented on src/main/java/com/bamdoliro/maru/application/form/UploadFormUseCase.java:
```diff
@@ -32,6 +37,12 @@ public UrlResponse execute(User user, FileMetadata fileMetadata) {
+            throw new OutOfApplicationPeriodException();
+        }
+    }
+
```
P3) 제출, 최종 제출, 수정, pdf 제출, 증명사진 제출에서 모두 중복된 검증 로직을 사용하는데 중복을 최소화할 방법은 없을까요?

cabbage16 replied:
AOP 사용해서 중복되는 로직 통합했습니다!

cabbage16 committed:
perf(#69): 원서 접수 기간 검증 AOP 적용
- 원서 접수 기간을 검증하는 중복 로직이 너무 많아서 aop를 이용해서 공통 로직을 통합했어요.
- 입학등록원 및 금연서약서 제출 기간 검증은 한군데에서밖에 사용되지 않아서 굳이 aop로 만들지 않았어요.

cabbage16 committed:
test(#69): 원서 접수 기간 검증 AOP 테스트 코드 작성
- 원서 접수 기간 검증을 AOP로 옮기면서 AOP 테스트를 만들었어요.
- 기존에 UseCaseTest에 있던 기간 검증 테스트 코드를 모두 삭제했어요.
